### PR TITLE
Examples: add contact shadow example

### PR DESCRIPTION
**Pull request by marcofugaro**
Following discussion in #18746

### CODE REVIEW

**Comment by looeee** [comment]
[Live version](https://raw.githack.com/marcofugaro/three.js/contact-shadow/examples/webgl_shadow_contact.html). 

I'm not seeing any shadows though.

Annotation 2020-03-01 084606

**Comment by marcofugaro** [comment]
Hmm @looeee which browser are you using?

**Comment by Mugen87** [comment]
I see the same like @looeee  on Chrome 80.0.3987.122 and Firefox 73.0.1,  tested on macOS 10.15.3.

File: examples/webgl_shadow_contact.html [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+
+			function init() {
+
+				camera = new THREE.PerspectiveCamera( 50, window.innerWidth / window.innerHeight, 0.1, 100 );
+				camera.position.set( 0.5, 1, 2 );
+
+				scene = new THREE.Scene();
+
+				stats = new Stats();
+				document.body.appendChild( stats.dom );
```
Comment (Mugen87) [inline examples/webgl_shadow_contact.html]: You have to update `stats` in the animation loop.

Comment (marcofugaro) [reply examples/webgl_shadow_contact.html]: Done!

**Comment by marcofugaro** [comment]
Alright fixed, not a browser issue, my fork was referring to an outdated version of the Three.js build.

Githack should update soon.

File: examples/webgl_shadow_contact.html [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+				renderTarget = new THREE.WebGLRenderTarget( 512, 512 );
+				renderTarget.texture.generateMipmaps = false;
+
+				// the render target that we will use to blur the first render target
+				renderTargetBlur = new THREE.WebGLRenderTarget( 512, 512 );
+				renderTargetBlur.texture.generateMipmaps = false;
+
+
+				// make a plane and make it face up
+				planeGeometry = new THREE.PlaneGeometry( PLANE_WIDTH, PLANE_HEIGHT ).rotateX( Math.PI / 2 );
```
Comment (Mugen87) [inline examples/webgl_shadow_contact.html]: Can you please use `PlaneBufferGeometry` here?

File: examples/webgl_shadow_contact.html [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+
+				// the camera to render the depth material from
+				shadowCamera = new THREE.OrthographicCamera( - PLANE_WIDTH / 2, PLANE_WIDTH / 2, PLANE_HEIGHT / 2, - PLANE_HEIGHT / 2, 0, CAMERA_HEIGHT );
+				shadowCamera.rotation.x = Math.PI / 2; // get the camera to look up
+				shadowGroup.add( shadowCamera );
+
+				cameraHelper = new THREE.CameraHelper( shadowCamera );
+
+				// like MeshDepthMaterial, but goes from black to transparent
+				depthMaterial = new THREE.ShaderMaterial( {
```
Comment (Mugen87) [inline examples/webgl_shadow_contact.html]: Could you also use `MeshDepthMaterial` here and perform the changes with `onBeforeCompile()`?

File: examples/webgl_shadow_contact.html [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+
+						scene.remove( cameraHelper );
+
+					}
+
+				} );
+
+				//
+
+				renderer = new THREE.WebGLRenderer( { antialias: true, alpha: true } );
```
Comment (Mugen87) [inline examples/webgl_shadow_contact.html]: I don't think it's necessary to set `alpha` to `true`...

File: examples/webgl_shadow_contact.html [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+				renderer.setPixelRatio( window.devicePixelRatio );
+				renderer.setSize( window.innerWidth, window.innerHeight );
+				renderer.outputEncoding = THREE.sRGBEncoding;
```
Comment (Mugen87) [inline examples/webgl_shadow_contact.html]: This line can also be removed.

File: examples/webgl_shadow_contact.html (hunk 4 above)
Comment (marcofugaro) [reply examples/webgl_shadow_contact.html]: Yes, it's to show that the shadows rendered are transparent (i was using them in AR).
The white you see is the white of the body.

Should I use `renderer.setClearColor(0xffffff, 1)` instead?

Comment (Mugen87) [reply examples/webgl_shadow_contact.html]: Try it with `scene.background = new THREE.Color( 0xffffff );`

**Comment by looeee** [comment]
I think this would look better if the ground plane was not transparent. It looks kind of weird when the object cut through their shadows:

Annotation 2020-03-01 212823

Or is there a reason why the ground is transparent?

**Comment by marcofugaro** [comment]
@looeee yeah the objective was to show that they can be rendered against a transparent plane, However I can try putting a slider with the floor opacity

**Comment by munrocket** [comment]
Please, add screenshot in PR with
```
npm run make-screenshot webgl_shadow_contact
```
we need to add automatic suggestion like this inside Travis

**Comment by marcofugaro** [comment]
Thank you @munrocket didn't know that. Also a "Contributing" section in the readme or in the docs would be nice

**Comment by munrocket** [comment]
Ok, will do.

**Comment by munrocket** [comment]
@marcofugaro actually we already have new contribution guide. And without screenshot this probably will not merged, because every new commit will be with broken Travis too. So try to make it, if it possible on your GPU.

**Comment by mrdoob** [comment]
> @marcofugaro actually we already have new contribution guide. And without screenshot this probably will not merged

Actually. Lets revisit that.

I would prefer not to make the contribution process more complicated. I can just run the make-screenshot myself.

**Comment by marcofugaro** [comment]
No problem, just had to get around to doing it 🙂 

@munrocket you're right, I found it in the wiki! I've been using Three for years and didn't know the wiki even existed!

Also I added a plane with the color as suggested by @looeee, ended up making another mesh rather than playing with blendings in the shadow plane. Please review @Mugen87

File: examples/webgl_shadow_contact.html [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+				verticalBlurMaterial.uniforms.v.value = amount * 1 / 256;
+
+				renderer.setRenderTarget( renderTarget );
+				renderer.render( blurPlane, shadowCamera );
+
+				blurPlane.visible = false;
+
+			}
+
+			function animate( time ) {
```
Comment (Mugen87) [inline examples/webgl_shadow_contact.html]: It seems `time` is defined but never used.

**Comment by munrocket** [comment]
Kek, travis froze on raycasting. Never happened before and here again 😏

**Comment by munrocket** [comment]
Looks like the current screenshot generator work only for me.
```
ERROR! Diff wrong in 0.987 of pixels in file: webgl_shadow_contact
```

**Comment by munrocket** [comment]
I need to ask help from friends with Windows machines and another GPUs.
@marcofugaro please, just add `webgl_shadow_contact` in [exception_list](https://github.com/mrdoob/three.js/blob/dev/test/e2e/puppeteer.js) if it's not hard for you.
Sorry but you was the first who generated a new screenshot by yourself.

**Comment by marcofugaro** [comment]
No problem, I'll leave you guys generate the screenshots as @mrdoob said 😅

**Review by Mugen87 — APPROVED** [review]

**Comment by munrocket** [comment]
@Mugen87 @looeee I asked two people to check current regression tests. When software is working _only_ on my machine and in CI this is beyond my force. I literally can't fix it even if I want.

Right now I have two version _one_ in three.js repo, and [_another_](https://github.com/munrocket/puppeteer-webgl-ci) experimental in my repo, the last change that I have is to try to mix it in draft PR with new approach and make a stress test on current codebase. 

Probably some flag from that repo will be suitable for Marcofugaro and Mr.Doob PCs.

**Comment by makc** [comment]
what a weird thing about this demo, it uses two blur shaders (h/v) each of them has a bunch of texture reads. normally this is slow on my old laptop, but this on is very fast. I wish I could understand why.

**Comment by marcofugaro** [comment]
@makc not even that, the demo is doing 2 blur rounds, so in total there are 4 passes each frame!
The blur shader implemented in Three.js is very efficient, you can read more about it [here](http://vodacek.zvb.cz/archiv/251.html) and [here](http://rastergrid.com/blog/2010/09/efficient-gaussian-blur-with-linear-sampling/)

**Comment by Mugen87** [comment]
There is now a merge conflict  because of #18803. Do you mind removing `webgl_shadow_contact.png` from your PR?

**Comment by marcofugaro** [comment]
Sure!

**Comment by marcofugaro** [comment]
@Mugen87 done!

**Comment by munrocket** [comment]
Sorry that I am talking about tests again here, but I am asked people with nvidia 2060 and 1060ti. And all of them able to generate new screenshots. @marcofugaro what videocard/OS do you use, I can't reproduce the bug.

**Comment by mrdoob** [comment]
Thanks!

**Pull request merged**